### PR TITLE
Add pytorchrl

### PR DESCRIPTION
**Pull request by jan-janssen**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by jan-janssen** [comment]
```
Encountered problems while solving:
  - nothing provides requested tensordict >=0.4.0
```

**Comment by jan-janssen** [comment]
Waiting for https://github.com/conda-forge/tensordict-feedstock/pull/11

**Comment by jan-janssen** [comment]
```
  ERROR (pytorchrl,lib/python3.12/site-packages/torchrl/_torchrl.so): $RPATH/libtorch_python.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
```

**Comment by jan-janssen** [comment]
@conda-forge/staged-recipes, ready for review

**Comment by jan-janssen** [comment]
@conda-forge/help-python-c

File: recipes/pytorchrl/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  url: https://github.com/pytorch/rl/archive/v{{ version }}.tar.gz
+  sha256: 8851a84316f2a1d61d23ec753c90b545bc6479890ec65c5278987ae7c8a2ceec
+
+build:
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+  skip: true  # [win]
+  missing_dso_whitelist:
+    - $RPATH/libtorch_python.so  # [linux]
+    - $RPATH/libtorch_python.dylib  # [osx]
```
Comment (ocefpaf) [inline recipes/pytorchrl/meta.yaml:18]:
```suggestion
    - $RPATH/libtorch_python$SHLIB_EXT  # [not win]
```

**Comment by jan-janssen** [comment]
@ocefpaf I tried your suggestion, but this causes the tests to fail: 
```
Warning: rpath /home/conda/staged-recipes/build_artifacts/pytorchrl_1719214381235/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/pytorchrl_1719214381235/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p (removing it)
   INFO: sysroot: '/home/conda/staged-recipes/build_artifacts/pytorchrl_1719214381235/_build_env/x86_64-conda-linux-gnu/sysroot/' files: '['usr/share/zoneinfo/zone1970.tab', 'usr/share/zoneinfo/zone.tab', 'usr/share/zoneinfo/tzdata.zi', 'usr/share/zoneinfo/right/Zulu']'
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO lib/libc10.so found in conda-forge/linux-64::libtorch==2.3.1=cpu_mkl_h0bb0d08_100
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO lib/libtorch_cpu.so found in conda-forge/linux-64::libtorch==2.3.1=cpu_mkl_h0bb0d08_100
  ERROR (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): $RPATH/libtorch_python.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO lib/libstdc++.so.6 found in conda-forge/linux-64::libstdcxx-ng==13.2.0=hc0a3c3a_13
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO lib/libgcc_s.so.1 found in conda-forge/linux-64::libgcc-ng==13.2.0=h77fa898_13
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.17=h4a8ded7_14
   INFO (pytorchrl,lib/python3.11/site-packages/torchrl/_torchrl.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/ld-linux-x86-64.so.2 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.17=h4a8ded7_14
WARNING (pytorchrl): interpreter (Python) package conda-forge/linux-64::python==3.11.9=hb806964_0_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (pytorchrl): run-exports library package conda-forge/linux-64::numpy==2.0.0=py311h1461c94_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (pytorchrl): run-exports library package conda-forge/linux-64::pytorch==2.3.1=cpu_mkl_py311hcb16b95_100 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
My suggestion would be to revert your suggestion and stay with the previous version. It seem like the usage of environment variables in the `meta.yaml` file was not successful.

**Comment by ocefpaf** [comment]
Sorry, it should've been `{{ SHLIB_EXT }}` instead. But it seems you already figured that out.

**Comment by jan-janssen** [comment]
@ocefpaf I included your recommendations and the tests passed - can you take another look at this pull request?

File: recipes/pytorchrl/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+source:
+  url: https://github.com/pytorch/rl/archive/v{{ version }}.tar.gz
+  sha256: 8851a84316f2a1d61d23ec753c90b545bc6479890ec65c5278987ae7c8a2ceec
+
+build:
+  script: {{ PYTHON }} -m pip install . -vv --no-deps --no-build-isolation
+  number: 0
+  skip: true  # [win]
+  missing_dso_whitelist:
+    - $RPATH/libtorch_python{{ SHLIB_EXT }}  # [not win]
```
Comment (ocefpaf) [inline recipes/pytorchrl/meta.yaml:17]: I'm not sure about this part. @isuruf or @jaimergp, sorry for the ping but it this OK to go?

Comment (jan-janssen) [reply recipes/pytorchrl/meta.yaml:17]: The same is used in a couple of packages on `conda-forge`:
* https://github.com/conda-forge/sinabs-exodus-feedstock
* https://github.com/conda-forge/theseus-ai-feedstock
* https://github.com/conda-forge/pytorch3d-feedstock

**Comment by jan-janssen** [comment]
@ocefpaf Any update on this pull request?

File: recipes/pytorchrl/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipes/pytorchrl/meta.yaml:17]: Waiting for someone with more knowledge on this to chime in.

Comment (isuruf) [inline recipes/pytorchrl/meta.yaml:17]:
```suggestion
  rpaths:
    - lib
    - {{ SP_DIR }}/torch
```

Comment (jan-janssen) [reply recipes/pytorchrl/meta.yaml]: @isuruf with this modification I again get the following error:
```
Warning: rpath /home/conda/staged-recipes/build_artifacts/pytorchrl_1719418355124/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/pytorchrl_1719418355124/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p (removing it)
   INFO: sysroot: '/home/conda/staged-recipes/build_artifacts/pytorchrl_1719418355124/_build_env/x86_64-conda-linux-gnu/sysroot/' files: '['usr/share/zoneinfo/zone1970.tab', 'usr/share/zoneinfo/zone.tab', 'usr/share/zoneinfo/tzdata.zi', 'usr/share/zoneinfo/right/Zulu']'
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO lib/libc10.so found in conda-forge/linux-64::libtorch==2.3.1=cpu_mkl_h0bb0d08_100
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO lib/libtorch_cpu.so found in conda-forge/linux-64::libtorch==2.3.1=cpu_mkl_h0bb0d08_100
  ERROR (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): $RPATH/libtorch_python.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO lib/libstdc++.so.6 found in conda-forge/linux-64::libstdcxx-ng==13.2.0=hc0a3c3a_13
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO lib/libgcc_s.so.1 found in conda-forge/linux-64::libgcc-ng==13.2.0=h77fa898_13
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.17=h4a8ded7_14
   INFO (pytorchrl,lib/python3.10/site-packages/torchrl/_torchrl.so): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/ld-linux-x86-64.so.2 found in CDT/compiler package conda-forge/noarch::sysroot_linux-64==2.17=h4a8ded7_14
WARNING (pytorchrl): run-exports library package conda-forge/linux-64::pytorch==2.3.1=cpu_mkl_py310h75865b9_100 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (pytorchrl): interpreter (Python) package conda-forge/linux-64::python==3.10.14=hd12c33a_0_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (pytorchrl): run-exports library package conda-forge/linux-64::numpy==2.0.0=py310h515e003_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

**Comment by jan-janssen** [comment]
@isuruf Thanks a lot - this fixed the issue. Can you give me a brief explanation why the default installation location is not suitable?

**Comment by isuruf** [comment]
Usually all the shared libraries live in `$PREFIX/lib` for conda-forge, however `llbitorch_python.so` is in `$SP_DIR/torch/lib`

**Comment by jan-janssen** [comment]
@ocefpaf and @isuruf Can this pull request be merged?

**Pull request merged**